File: app/src/main/java/choruscoldchain/app/VerifyOtpActivity.java
package choruscoldchain.app;

import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.material.button.MaterialButton;
import android.widget.TextView;
import java.io.IOException;
import java.io.InputStream;

public class VerifyOtpActivity extends AppCompatActivity {
    private static final String TAG = "VerifyOtpActivity";
    
    private EditText etOtp1, etOtp2, etOtp3, etOtp4;
    private MaterialButton btnVerify, btnBackToLogin;
    private TextView tvResendOtp;
    private ImageView ivChorusLogo;
    
    private AuthManager authManager;
    private String email, hashedUser;
    private boolean fromForgotPassword;
    private CountDownTimer timer;
    private int timeLeft = 60; // 60 seconds
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_verify_otp);
        
        // Set status bar text color to black
        getWindow().getDecorView().setSystemUiVisibility(
            getWindow().getDecorView().getSystemUiVisibility() | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
        );
        
        // Get data from intent
        Intent intent = getIntent();
        email = intent.getStringExtra("email");
        hashedUser = intent.getStringExtra("hashedUser");
        fromForgotPassword = intent.getBooleanExtra("fromForgotPassword", false);
        
        if (email == null || hashedUser == null) {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show();
            finish();
            return;
        }
        
        authManager = AuthManager.getInstance(this);
        
        initViews();
        setupListeners();
        loadChorusLogo();
        startTimer();
        
        // Auto focus first input
        etOtp1.requestFocus();
    }
    
    private void initViews() {
        etOtp1 = findViewById(R.id.etOtp1);
        etOtp2 = findViewById(R.id.etOtp2);
        etOtp3 = findViewById(R.id.etOtp3);
        etOtp4 = findViewById(R.id.etOtp4);
        btnVerify = findViewById(R.id.btnVerify);
        btnBackToLogin = findViewById(R.id.btnBackToLogin);
        tvResendOtp = findViewById(R.id.tvResendOtp);
        ivChorusLogo = findViewById(R.id.ivChorusLogo);
    }
    
    private void loadChorusLogo() {
        // Try to load chorus logo from assets in order of preference
        String[] logoPaths = {
            "images/chorus.png",           // Main chorus logo
            "chorus_logo.png",             // Alternative chorus logo
            "images/chorusWhite.jpeg"      // White version if needed
        };
        
        for (String logoPath : logoPaths) {
            try {
                InputStream inputStream = getAssets().open(logoPath);
                Drawable drawable = Drawable.createFromStream(inputStream, null);
                ivChorusLogo.setImageDrawable(drawable);
                inputStream.close();
                Log.d(TAG, "Successfully loaded chorus logo from: " + logoPath);
                return; // Successfully loaded, exit the method
            } catch (IOException e) {
                Log.d(TAG, "Could not load chorus logo from: " + logoPath);
                // Continue to next option
            }
        }
        
        // Fallback to drawable if all asset loading fails
        Log.w(TAG, "Falling back to drawable chorus logo");
        ivChorusLogo.setImageResource(R.drawable.ic_chorus_logo);
    }
    
    private void setupListeners() {
        btnVerify.setOnClickListener(v -> handleVerify());
        btnBackToLogin.setOnClickListener(v -> handleBack());
        tvResendOtp.setOnClickListener(v -> handleResendOtp());
        
        // Setup OTP input listeners for auto-focus
        setupOtpInputListeners();
    }
    
    private void setupOtpInputListeners() {
        TextWatcher otpWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}
            
            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 1) {
                    // Move to next input
                    if (etOtp1.hasFocus()) {
                        etOtp2.requestFocus();
                    } else if (etOtp2.hasFocus()) {
                        etOtp3.requestFocus();
                    } else if (etOtp3.hasFocus()) {
                        etOtp4.requestFocus();
                    }
                } else if (s.length() == 0) {
                    // Move to previous input on backspace
                    if (etOtp2.hasFocus()) {
                        etOtp1.requestFocus();
                    } else if (etOtp3.hasFocus()) {
                        etOtp2.requestFocus();
                    } else if (etOtp4.hasFocus()) {
                        etOtp3.requestFocus();
                    }
                }
            }
        };
        
        etOtp1.addTextChangedListener(otpWatcher);
        etOtp2.addTextChangedListener(otpWatcher);
        etOtp3.addTextChangedListener(otpWatcher);
        etOtp4.addTextChangedListener(otpWatcher);
    }
    
    private void handleVerify() {
        String otp1 = etOtp1.getText().toString();
        String otp2 = etOtp2.getText().toString();
        String otp3 = etOtp3.getText().toString();
        String otp4 = etOtp4.getText().toString();
        
        String fullOtp = otp1 + otp2 + otp3 + otp4;
        
        if (fullOtp.length() != 4) {
            Toast.makeText(this, "Please enter the complete 4-digit OTP.", Toast.LENGTH_LONG).show();
            return;
        }
        
        btnVerify.setEnabled(false);
        btnVerify.setText("Loading...");
        
        authManager.verifyOtp(hashedUser.toLowerCase(), fullOtp, new AuthManager.AuthCallback<AuthModels.VerifyOtpResponse>() {
            @Override
            public void onSuccess(AuthModels.VerifyOtpResponse result) {
                runOnUiThread(() -> {
                    btnVerify.setEnabled(true);
                    btnVerify.setText("Verify");
                    
                    if (result.isSuccess()) {
                        // Save the access token from OTP verification
                        String accessToken = result.getAccessToken();
                        if (accessToken != null) {
                            // For OTP verification, we don't have role info yet, so save without role
                            authManager.saveAuthData(accessToken, email);
                        }
                        
                        Toast.makeText(VerifyOtpActivity.this, "OTP Verified!", Toast.LENGTH_SHORT).show();
                        navigateToResetPassword();
                    } else {
                        Toast.makeText(VerifyOtpActivity.this, 
                            "Invalid OTP. Please try again.", Toast.LENGTH_LONG).show();
                    }
                });
            }
            
            @Override
            public void onError(String error) {
                runOnUiThread(() -> {
                    btnVerify.setEnabled(true);
                    btnVerify.setText("Verify");
                    Toast.makeText(VerifyOtpActivity.this, 
                        "Something went wrong. Please try again.", Toast.LENGTH_LONG).show();
                });
            }
        });
    }
    
    private void handleResendOtp() {
        if (timeLeft > 0) return;
        
        btnVerify.setEnabled(false);
        btnVerify.setText("Loading...");
        
        authManager.forgotPassword(email, new AuthManager.AuthCallback<AuthModels.ForgotPasswordResponse>() {
            @Override
            public void onSuccess(AuthModels.ForgotPasswordResponse result) {
                runOnUiThread(() -> {
                    btnVerify.setEnabled(true);
                    btnVerify.setText("Verify");
                    
                    if (result.isSuccess()) {
                        Toast.makeText(VerifyOtpActivity.this, "OTP has been sent to your email.", Toast.LENGTH_LONG).show();
                        hashedUser = result.getHashedUser();
                        clearOtpFields();
                        startTimer();
                        etOtp1.requestFocus();
                    } else {
                        Toast.makeText(VerifyOtpActivity.this, 
                            "Failed to generate OTP. Please try again.", Toast.LENGTH_LONG).show();
                    }
                });
            }
            
            @Override
            public void onError(String error) {
                runOnUiThread(() -> {
                    btnVerify.setEnabled(true);
                    btnVerify.setText("Verify");
                    Toast.makeText(VerifyOtpActivity.this, 
                        "Something went wrong. Please try again.", Toast.LENGTH_LONG).show();
                });
            }
        });
    }
    
    private void handleBack() {
        finish();
    }
    
    private void startTimer() {
        timeLeft = 60;
        tvResendOtp.setText("Resend OTP in (60s)");
        tvResendOtp.setClickable(false);
        
        if (timer != null) {
            timer.cancel();
        }
        
        timer = new CountDownTimer(60000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                timeLeft = (int) (millisUntilFinished / 1000);
                tvResendOtp.setText("Resend OTP in (" + timeLeft + "s)");
            }
            
            @Override
            public void onFinish() {
                timeLeft = 0;
                tvResendOtp.setText("Resend OTP");
                tvResendOtp.setClickable(true);
            }
        }.start();
    }
    
    private void clearOtpFields() {
        etOtp1.setText("");
        etOtp2.setText("");
        etOtp3.setText("");
        etOtp4.setText("");
    }
    
    private void navigateToResetPassword() {
        Intent intent = new Intent(this, ResetPasswordActivity.class);
        intent.putExtra("email", email);
        startActivity(intent);
        finish();
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (timer != null) {
            timer.cancel();
        }
    }
} 

File: app/src/main/java/com/choruscoldchain/ui/activities/VerifyOtpActivity.kt
package com.choruscoldchain.ui.activities

import android.os.Bundle
import android.os.CountDownTimer
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.button.MaterialButton
import com.choruscoldchain.R
import com.choruscoldchain.auth.AuthManager
import com.choruscoldchain.auth.AuthModels
import com.choruscoldchain.utils.InstallationIdManager

class VerifyOtpActivity : AppCompatActivity() {
    private lateinit var etOtp1: EditText
    private lateinit var etOtp2: EditText
    private lateinit var etOtp3: EditText
    private lateinit var etOtp4: EditText
    private lateinit var btnVerify: MaterialButton
    private lateinit var btnBackToLogin: MaterialButton
    private lateinit var tvResendOtp: TextView
    private lateinit var ivChorusLogo: ImageView

    private lateinit var authManager: AuthManager
    private lateinit var email: String
    private lateinit var hashedUser: String
    private var fromForgotPassword: Boolean = false
    private var timer: CountDownTimer? = null
    private var timeLeft = 60

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_verify_otp)

        window.decorView.systemUiVisibility =
            window.decorView.systemUiVisibility or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR

        email = intent.getStringExtra("email") ?: run {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show()
            finish(); return
        }
        hashedUser = intent.getStringExtra("hashedUser") ?: run {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show()
            finish(); return
        }
        fromForgotPassword = intent.getBooleanExtra("fromForgotPassword", false)

        authManager = AuthManager.getInstance(this)

        initViews()
        setupListeners()
        setupAppId()
        startTimer()
        etOtp1.requestFocus()
    }

    private fun initViews() {
        etOtp1 = findViewById(R.id.etOtp1)
        etOtp2 = findViewById(R.id.etOtp2)
        etOtp3 = findViewById(R.id.etOtp3)
        etOtp4 = findViewById(R.id.etOtp4)
        btnVerify = findViewById(R.id.btnVerify)
        btnBackToLogin = findViewById(R.id.btnBackToLogin)
        tvResendOtp = findViewById(R.id.tvResendOtp)
        ivChorusLogo = findViewById(R.id.ivChorusLogo)
    }

    private fun setupAppId() {
        val tvAppId = findViewById<TextView>(R.id.tv_scan_app_id)
        tvAppId?.text = "App ID: ${InstallationIdManager.getInstance(this).getInstallationId()}"
    }

    private fun setupListeners() {
        btnVerify.setOnClickListener { handleVerify() }
        btnBackToLogin.setOnClickListener { finish() }
        tvResendOtp.setOnClickListener { handleResendOtp() }

        val watcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                if (s?.length == 1) {
                    when {
                        etOtp1.hasFocus() -> etOtp2.requestFocus()
                        etOtp2.hasFocus() -> etOtp3.requestFocus()
                        etOtp3.hasFocus() -> etOtp4.requestFocus()
                    }
                } else if (s?.isEmpty() == true) {
                    when {
                        etOtp2.hasFocus() -> etOtp1.requestFocus()
                        etOtp3.hasFocus() -> etOtp2.requestFocus()
                        etOtp4.hasFocus() -> etOtp3.requestFocus()
                    }
                }
            }
        }

        etOtp1.addTextChangedListener(watcher)
        etOtp2.addTextChangedListener(watcher)
        etOtp3.addTextChangedListener(watcher)
        etOtp4.addTextChangedListener(watcher)
    }

    private fun handleVerify() {
        val fullOtp = listOf(etOtp1, etOtp2, etOtp3, etOtp4).joinToString("") { it.text.toString() }
        if (fullOtp.length != 4) {
            Toast.makeText(this, "Please enter the complete 4-digit OTP.", Toast.LENGTH_LONG).show()
            return
        }
        btnVerify.isEnabled = false
        btnVerify.text = "Loading..."

        authManager.verifyOtp(hashedUser.lowercase(), fullOtp, object : AuthManager.AuthCallback<AuthModels.VerifyOtpResponse> {
            override fun onSuccess(result: AuthModels.VerifyOtpResponse) {
                runOnUiThread {
                    btnVerify.isEnabled = true
                    btnVerify.text = "Verify"
                    if (result.isSuccess()) {
                        val accessToken = result.getAccessToken()
                        if (!accessToken.isNullOrEmpty()) {
                            authManager.saveAuthData(accessToken, email, null)
                        }
                        Toast.makeText(this@VerifyOtpActivity, "OTP Verified!", Toast.LENGTH_SHORT).show()
                        navigateToResetPassword()
                    } else {
                        Toast.makeText(this@VerifyOtpActivity, "Invalid OTP. Please try again.", Toast.LENGTH_LONG).show()
                    }
                }
            }

            override fun onError(error: String) {
                runOnUiThread {
                    btnVerify.isEnabled = true
                    btnVerify.text = "Verify"
                    Toast.makeText(this@VerifyOtpActivity, "Something went wrong. Please try again.", Toast.LENGTH_LONG).show()
                }
            }
        })
    }

    private fun handleResendOtp() {
        if (timeLeft > 0) return
        btnVerify.isEnabled = false
        btnVerify.text = "Loading..."
        authManager.forgotPassword(email, object : AuthManager.AuthCallback<AuthModels.ForgotPasswordResponse> {
            override fun onSuccess(result: AuthModels.ForgotPasswordResponse) {
                runOnUiThread {
                    btnVerify.isEnabled = true
                    btnVerify.text = "Verify"
                    if (result.isSuccess()) {
                        Toast.makeText(this@VerifyOtpActivity, "OTP has been sent to your email.", Toast.LENGTH_LONG).show()
                        hashedUser = result.hashedUser ?: hashedUser
                        clearOtpFields()
                        startTimer()
                        etOtp1.requestFocus()
                    } else {
                        Toast.makeText(this@VerifyOtpActivity, "Failed to generate OTP. Please try again.", Toast.LENGTH_LONG).show()
                    }
                }
            }

            override fun onError(error: String) {
                runOnUiThread {
                    btnVerify.isEnabled = true
                    btnVerify.text = "Verify"
                    Toast.makeText(this@VerifyOtpActivity, "Something went wrong. Please try again.", Toast.LENGTH_LONG).show()
                }
            }
        })
    }

    private fun startTimer() {
        timeLeft = 60
        tvResendOtp.text = "Resend OTP in (60s)"
        tvResendOtp.isClickable = false
        timer?.cancel()
        timer = object : CountDownTimer(60000, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                timeLeft = (millisUntilFinished / 1000).toInt()
                tvResendOtp.text = "Resend OTP in (${timeLeft}s)"
            }
            override fun onFinish() {
                timeLeft = 0
                tvResendOtp.text = "Resend OTP"
                tvResendOtp.isClickable = true
            }
        }.start()
    }

    private fun clearOtpFields() {
        etOtp1.setText("")
        etOtp2.setText("")
        etOtp3.setText("")
        etOtp4.setText("")
    }

    private fun navigateToResetPassword() {
        val intent = android.content.Intent(this, ResetPasswordActivity::class.java)
        intent.putExtra("email", email)
        intent.putExtra("accessToken", authManager.getToken())
        startActivity(intent)
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        timer?.cancel()
    }
}


